packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -1796,6 +1831,48 @@ function handleThrow(root, thrownValue): void {
   }
 }
 
+function shouldAttemptToSuspendUntilDataResolves() {
+  // TODO: We should be able to move the
+  // renderDidSuspend/renderDidSuspendWithDelay logic into this function,
+  // instead of repeating it in the complete phase. Or something to that effect.
+
+  if (includesOnlyRetries(workInProgressRootRenderLanes)) {
+    // We can always wait during a retry.
+    return true;
+  }
+
+  // TODO: We should be able to remove the equivalent check in
+  // finishConcurrentRender, and rely just on this one.
+  if (includesOnlyTransitions(workInProgressRootRenderLanes)) {
+    const suspenseHandler = getSuspenseHandler();
+    if (suspenseHandler !== null && suspenseHandler.tag === SuspenseComponent) {
+      const currentSuspenseHandler = suspenseHandler.alternate;
+      const nextProps: SuspenseProps = suspenseHandler.memoizedProps;
+      if (isBadSuspenseFallback(currentSuspenseHandler, nextProps)) {
```

Comment:
Don't we know this based on the Suspense context?

Reply:
The context stores the nearest "active" handler but it doesn't current track whether it's an avoided boundary, though it could. But that's effectively what this check does. It also checks if it's inside an offscreen tree:

https://github.com/facebook/react/blob/6883d79445f38e2cdabf6fe1fa04cb44c2644c1c/packages/react-reconciler/src/ReactFiberSuspenseContext.new.js#L61-L86